modules/objdetect/include/opencv2/objdetect/aruco_board.hpp

```diff
@@ -133,14 +133,20 @@ class CV_EXPORTS_W_SIMPLE CharucoBoard : public Board {
      * @param markerLength marker side length (same unit than squareLength)
      * @param dictionary dictionary of markers indicating the type of markers
      * @param ids array of id used markers
+     * @param legacy legacy chessboard pattern (pre 4.6.0). Pattern generation was changed for even row patterns (see https://github.com/opencv/opencv_contrib/issues/3291).
      * The first markers in the dictionary are used to fill the white chessboard squares.
      */
+    CV_WRAP CharucoBoard(const Size& size, float squareLength, float markerLength,
```

Comment:
I agree that we should not have an overload ctor, but rather a legacy flag (**set to true by default**) as the last parameter in the existing ctor (with an explanation in the ctor documentation), and go without a setLegacy method.

The reasons are below:
* Migrating from pre-4.6.0 to 4.6.0 breaks the code silently (when using a physical target manufactured with pre 4.6.0) and it is difficult for a novice developer to figure out why.
* Migrating from pre-4.5.5 to 4.7.0 breaks the code explicitly because the ctor is different (Size instead of two ints), as in @stefan523's example code in #23152. When the developer fixes the compilation error, they trigger the silent failures of the item above. If we introduce a legacy flag set to true by default, this is avoided and backward compatibility is preserved even if the developer is not aware of the legacy flag.
* Migrating from 4.6.0 to 4.7.0 requires the developer to explicitly update the ctor both in the Size parameters and in the legacy flag set to false (that is, assuming they are analyzing a target fabricated using 4.6.0 code). This is the only case that actually leads to backward compatibility issues per my suggestion, when the developer is not aware of the role of the legacy flag.

I think that it is more important to maintain backward compatibility with the dozens of released versions that have been employed to produce physical charuco targets that are still is use in production lines. Deciding otherwise will inevitably lead to systems version upgrades silently failing without an evident explanation.

Reply:
> I propose to not touch the constructor, but add `setLegacyFlag()` method. It's easier to remove legacy method in future than break compatibility for all code that uses Charuco.

I'm not sure if I understand the ask correctly.
I did keep the existing ctor as-is and added an overload with legacy flag in order to pass the ABI compliance check:
https://pullrequest.opencv.org/buildbot/builders/precommit_linux64/builds/102489/steps/Compare%20ABI%20dumps/logs/report-html
Adding a `setLegacyFlag()` method would add a new sympol as well. And this would break code, trigger an ABI compliance failure if removed in the future, would it not? I don't see the difference in that respect.

The existing ctor already creates a board. Adding a `setLegacyFlag()` method that is called after the constructor has run would mean to clear the board and recreate it differently. Is that really the ask? I don't see the advantage, only things that could go bad in the workflow. Or do I get it wrong?

Either way I'm fine with anything that enables usage of existing targets and data. @asmorkalov, please advise and I'll try to adapt.

Reply:
I believe we should minimize or do not bring code with legacy/buggy behavior from `opencv_contrib` into the main repository (as we don't really want to support/maintain that).

If `setLegacyFlag()` doesn't work, then I propose to create `CharucoBoardLegacy` or `cv::aruco::legacy::CharucoBoard` in `opencv_contrib/modules/aruco` for this purpose.
There is inheritance with `Board` base class and Pimpl design, but some trick is required to expose `Board::Impl` to `opencv_contrib` (we don't need that in public OpenCV headers).

Reply:
> Deciding otherwise will inevitably lead to systems version upgrades silently failing without an evident explanation.

one way or another, mentioning this change in the release notes would have been nice..

Reply:
> I believe we should minimize or do not bring code with legacy/buggy behavior from `opencv_contrib` into the main repository (as we don't really want to support/maintain that).
> 
> If `setLegacyFlag()` doesn't work, then I propose to create `CharucoBoardLegacy` or `cv::aruco::legacy::CharucoBoard` in `opencv_contrib/modules/aruco` for this purpose. There is inheritance with `Board` base class and Pimpl design, but some trick is required to expose `Board::Impl` to `opencv_contrib` (we don't need that in public OpenCV headers).

I don't mean that `setLegacyFlag()` does not work - just that it involves a bit more code changes and is not as elegant imho. But if this is the preferred option over adding a flag to the constructor, then I'll try to make a proposal. I'd like to avoid maintaining something in opencv_contrib in parallel.

The old behavior is not "buggy", it's just a slightly different board design, a variant. I still don't get why the board design had to change in the first place. There is no advantage, no reuse of existing code or something like that, but that's a different discussion. Now there are two versions out in the world.  Maybe "legacy" is the wrong word for it. Maybe calling it "even"/"odd", or "A"/"B" order. Or "WBF" (white-box-first). Does not matter much as long as it works.

I'll make a proposal for a setLegacyFlag() implementation.

Reply:
The insistence on the black-box-first design is admittedly baffling. I've had a custom WBF aruco slide made for microscopy calibration with chrome deposition on glass that cost hundreds of dollars, and now I can't upgrade opencv beyond 4.5.5 without my calibration code breaking completely.

Reply:
probably it makes sense to dub the BBF order (i.e. 4.6, 4.7 behavior) "legacy" and hence set the flag to false by default.

I doubt that with that many WBF patterns in the wild and WBF effectively being the default from 4.8 on BBF will be anything but legacy.

Reply:
Sorry, there is useful information for discussion here. Please don't mark as resolved for now.

Reply:
I'm interest as well to understand when this pul request will be close?